credentials/src/main/kotlin/web5/sdk/credentials/util/JwtUtil.kt

```diff
@@ -0,0 +1,131 @@
+   *
+   * Example:
+   * ```
+   * val signedVc = verifiableCredential.sign(myDid)
+   * ```
+   */
+  public fun sign(did: Did, assertionMethodId: String?, jwtPayload: JWTClaimsSet): String {
+    val didResolutionResult = DidResolvers.resolve(did.uri)
+    val assertionMethod = didResolutionResult.didDocument.findAssertionMethodById(assertionMethodId)
+
+    // TODO: ensure that publicKeyJwk is not null
+    val publicKeyJwk = JWK.parse(assertionMethod.publicKeyJwk)
+    val keyAlias = did.keyManager.getDeterministicAlias(publicKeyJwk)
+
+    // TODO: figure out how to make more reliable since algorithm is technically not a required property of a JWK
+    val algorithm = publicKeyJwk.algorithm
+    val jwsAlgorithm = JWSAlgorithm.parse(algorithm.toString())
+
+    val kid = when (assertionMethod.id.isAbsolute) {
+      true -> assertionMethod.id.toString()
+      false -> "${did.uri}${assertionMethod.id}"
+    }
+
+    val jwtHeader = JWSHeader.Builder(jwsAlgorithm)
+      .type(JOSEObjectType.JWT)
+      .keyID(kid)
+      .build()
+
+    val jwtObject = SignedJWT(jwtHeader, jwtPayload)
+    val toSign = jwtObject.signingInput
+    val signatureBytes = did.keyManager.sign(keyAlias, toSign)
+
+    val base64UrlEncodedHeader = jwtHeader.toBase64URL()
+    val base64UrlEncodedPayload = jwtPayload.toPayload().toBase64URL()
+    val base64UrlEncodedSignature = Base64URL(Convert(signatureBytes).toBase64Url(padding = false))
+
+    return "$base64UrlEncodedHeader.$base64UrlEncodedPayload.$base64UrlEncodedSignature"
+  }
+
+  /**
+   * Verifies the integrity and authenticity of a JSON Web Token (JWT).
+   *
+   * This function performs several crucial validation steps to ensure the trustworthiness of the provided VC:
+   * - Parses and validates the structure of the JWT.
+   * - Ensures the presence of critical header elements `alg` and `kid` in the JWT header.
+   * - Resolves the Decentralized Identifier (DID) and retrieves the associated DID Document.
+   * - Validates the DID and establishes a set of valid verification method IDs.
+   * - Identifies the correct Verification Method from the DID Document based on the `kid` parameter.
+   * - Verifies the JWT's signature using the public key associated with the Verification Method.
+   *
+   * If any of these steps fail, the function will throw a [SignatureException] with a message indicating the nature of the failure.
+   *
+   * @param jwtString The JWT as a [String].
+   * @throws SignatureException if the verification fails at any step, providing a message with failure details.
+   * @throws IllegalArgumentException if critical JWT header elements are absent.
+   */
+  public fun verify(jwtString: String) {
+    val jwt = JWTParser.parse(jwtString) as SignedJWT // validates JWT
+
+    require(jwt.header.algorithm != null && jwt.header.keyID != null) {
+      "Signature verification failed: Expected JWS header to contain alg and kid"
+    }
+
+    val verificationMethodId = jwt.header.keyID
```

Comment:
noting that these can be different

for example we may have a key id as a JWK thumbprint but a verification method ID as `did:example:abcd#key-1`